io/net/net_tools.py

```diff
@@ -0,0 +1,374 @@
+
+        self.lo_intf_state = "Up"
+        if "lo:1" in process.system_output("ifconfig", verbose=False):
+            self.skip("alias for an loopback interface is configured")
+        self.alias = None
+        install_dependencies()
+        self.ipv6 = False
+        if os.path.exists("/proc/net/if_inet6"):
+            self.ipv6 = True
+
+    def test_ifconfig(self):
+        """
+        Verify the functionality of the ifconfig
+        """
+        output = process.run("ifconfig", verbose=False,
+                             env={"LANG": "C"})
+        if "Local Loopback" not in output.stdout:
+            self.error("unexpected output of ifconfig")
+        if self.ipv6:
+            if "inet6" not in output.stdout:
+                self.error("Did not see IPV6 info")
+        # setup and verify an alias interface
+        self.alias = "lo:1"
+        process.system("ifconfig lo:1 127.0.0.240  netmask 255.0.0.0",
+                       verbose=False,
+                       sudo=True)
+        output = process.run("ifconfig", verbose=False,
+                             env={"LANG": "C"})
+        if "lo:1" not in output.stdout:
+            self.error("Failed to configure alias for loopback")
+
+        ret = self._remove_alias(self.alias)
+        if ret:
+            self.error(ret)
+        self.alias = None
+        # Test to make loopback interface up and down
+        process.system("ifconfig lo down", verbose=False,
+                       sudo=True)
+        self.lo_intf_state = "Down"
+        ret = process.SubProcess("ping -c 5 -w 5 localhost",
+                                 verbose=False).run()
+        if "100% packet loss" not in ret.stdout:
+            self.error("loopback interface is not in down")
+
+        ret = self._restore_lo_intf_state()
+        if ret:
+            self.error(ret)
+        self.lo_intf_state = "Up"
+
+    @staticmethod
+    def _remove_alias(name):
+        process.system("ifconfig %s down" % name,
+                       verbose=False,
+                       sudo=True)
+        output = process.run("ifconfig", verbose=False,
+                             env={"LANG": "C"})
+        if "lo:1" in output.stdout:
+            return "Failed to remove alias for loopback"
+
+    @staticmethod
+    def _restore_lo_intf_state():
+        process.system("ifconfig lo up", verbose=False,
+                       sudo=True)
+        output = process.system_output("ping -c 5 -w 5 localhost",
+                                       verbose=False,
+                                       env={"LANG": "C"})
+        if "0% packet loss" not in output:
+            return "loopback intf is not coming up"
+
+    def tearDown(self):
+        if self.alias:
+            ret = self._remove_alias(self.alias)
+            if ret:
+                self.error(ret)
+        if self.lo_intf_state is not "Up":
+            ret = self._restore_lo_intf_state()
+            if ret:
+                self.error(ret)
```

Comment:
The `tearDown` has to always complete all the phases and raise the exception at the end. So how about:

```python
errs = []
if self.alias:
    try:
        self._remove_alias(self.alias)
    except Exception:
        errs.append("remove alias")
if self.lo_intf_state is not "Up":
    try:
        self._restore_lo_intf_state()
    except Exception:
        errs.append("restore lo state")
if errs:
    self.error("Failed to: %s" % ", ".join(errs))